File: golang/array_string/insert_delete_random.go
package main

import (
	"math/rand"
	"time"
)

type RandomizedSet struct {
	values  []int
	indices map[int]int
}

func Constructor() RandomizedSet {
	rand.NewSource(time.Now().UnixNano())

	return RandomizedSet{
		values:  []int{},
		indices: map[int]int{},
	}
}

func (rs *RandomizedSet) Insert(val int) bool {
	if _, exists := rs.indices[val]; exists {
		return false
	}

	rs.values = append(rs.values, val)

	rs.indices[val] = len(rs.values) - 1

	return true
}

func (rs *RandomizedSet) Remove(val int) bool {
	index, exists := rs.indices[val]

	if !exists {
		return false
	}

	lastVal := rs.values[len(rs.values)-1]

	rs.values[index] = lastVal

	rs.indices[lastVal] = index

	rs.values = rs.values[:len(rs.values)-1]

	delete(rs.indices, val)

	return true
}

func (rs *RandomizedSet) GetRandom() int {
	return rs.values[rand.Intn(len(rs.values))]
}

File: python/array_string/insert_delete_random.py
import random


class RandomizedSet:
    def __init__(self) -> None:
        self.values = []
        self.indices = {}

    def insert(self, val: int) -> bool:
        if val in self.indices:
            return False

        self.indices[val] = len(self.values)

        self.values.append(val)

        return True

    def remove(self, val: int) -> bool:
        if val not in self.values:
            return False

        last_val = self.values[-1]
        index_to_remove = self.indices[val]

        self.values[index_to_remove] = last_val
        self.indices[last_val] = index_to_remove

        self.values.pop()

        del self.indices[val]

        return True

    def getRandom(self) -> int:
        return random.choice(self.values)
